kernel/src/hil/block_storage.rs

```diff
@@ -0,0 +1,268 @@
+//!
+//! Operates on blocks, with an optional way to read unaligned bytes.
+//!
+//! Here's an example implementation for a raw flash chip:
+//!
+//! ```rust
+//! use kernel::hil;
+//! use kernel::ErrorCode;
+//!
+//! const WRITE_BLOCK_BYTES: u32 = 256;
+//! const ERASE_BLOCK_BYTES: u32 = 4096;
+//!
+//! struct RawFlashChip {};
+//!
+//! type WriteRegion = hil::block_storage::Region<WRITE_BLOCK_BYTES>;
+//! type EraseRegion = hil::block_storage::Region<ERASE_BLOCK_BYTES>;
+//!
+//! impl hil::block_storage::BlockStorage<WRITE_BLOCK_BYTES, ERASE_BLOCK_BYTES> for RawFlashChip {
+//!     (implement associated functions here)
+//! }
+//! ```
+//!
+//! Similar implementation for a flash chip with FTL, or for a magnetic drive:
+//!
+//! ```rust
+//! use kernel::hil;
+//! use kernel::ErrorCode;
+//!
+//! const BLOCK_BYTES: u32 = 4096;
+//!
+//! struct SimpleBlockDevice {};
+//!
+//! type Region = hil::block_storage::Region<BLOCK_BYTES>;
+//!
+//! impl hil::block_storage::BlockStorage<BLOCK_BYTES, BLOCK_BYTES> for SimpleBlockDevice {
+//!     // FTL takes care of multiple writes to the same block on flash,
+//!     // and magnetic drives don't need to erase at all.
+//!     fn erase(&self, region: &hil::block_storage::Region)
+//!         -> Result<(), ErrorCode>
+//!     { Ok(()) }
+//!     
+//!     (implement remaining associated functions here)
+//! }
+//! ```
+use crate::ErrorCode;
+
+/// An index to a block within device composed of `S`-sized blocks.
+#[derive(Clone, Copy)]
+pub struct BlockIndex<const S: u32>(pub u32);
+
+impl<const S: u32> BlockIndex<S> {
+    /// Returns the index that contains the address.
+    pub fn new_containing(address: u64) -> Self {
+        Self((address / S as u64) as u32)
+    }
+
+    /// Returns the index starting at the given address, if any.
+    pub fn new_starting_at(address: u64) -> Option<Self> {
+        if address % S as u64 == 0 {
+            Some(Self::new_containing(address))
+        } else {
+            None
+        }
+    }
+}
+
+impl<const S: u32> From<BlockIndex<S>> for u64 {
+    fn from(index: BlockIndex<S>) -> Self {
+        index.0 as u64 * S as u64
+    }
+}
+
+/// A memory region composed of consecutive `S`-sized blocks.
+#[derive(Clone, Copy)]
+pub struct Region<const S: u32> {
+    pub index: BlockIndex<S>,
+    pub length_blocks: u32,
+}
+
+/// Divide, round up
+fn div_ceil(a: u64, b: u64) -> u64 {
+    a / b // FIXME: trololo
+}
+
+impl<const S: u32> Region<S> {
+    /// Returns the smallest region made of blocks which contains
+    /// the specified range of addresses.
+    pub fn new_containing(range: AddressRange) -> Self {
+        let index = BlockIndex::new_containing(range.start_address);
+        Self {
+            index,
+            length_blocks: {
+                div_ceil(
+                    range.start_address % S as u64 + range.length_bytes as u64,
+                    S as u64,
+                ) as u32
+                // CAUTION: div_ceil is nightly
```

Comment:
what does this refer to?

Reply:
It's the `div_ceil` function which is needed here: https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil . This is also what the FIXME is referring to; I haven't found the right incantation to do the division yet.

Reply:
Ok makes sense, I was just expecting to see a feature.